docs/type_inference.md

```diff
@@ -26,12 +26,21 @@ Both
 - **verifiers**: implemented by `Op::verify()`, and
 - **shape functions**: implemented by `InferTypeOpInterface` like `Op::inferReturnTypes()` or `Op::inferReturnTypeComponents`
 
-may have verification code to check operands/attributes/results. An ideal split would be that: let the verifiers check the operands/attributes, then let shape functions to calculate inferred result types and check the compatibility against the real result types. However, in reality this split has a few problems:
+may have verification code to check operands/attributes/results. An initial split would be that: let the verifiers check the operands/attributes, then let shape functions only calculate inferred result types and check the compatibility against the real result types. However, in reality this split has a few problems:
 
-1. Duplicated code: for example in verifiers we do some processing on the operands then verify some intermediate results, then in shape functions these intermediate results are useful to infer the final results. These intermediate results have to be calculated twice.
-2. Maintenance burden: as verifications of an op are contained in two different methods.
+- The shape function can be called by the auto-gen `builder()` alone, without calling the verifier first. So the related inputs must be verified in shape function as well.
+- Duplicated code: for example in verifiers we do some processing on the operands then verify some intermediate results, then in shape functions these intermediate results are useful to infer the final results. These intermediate results have to be calculated twice.
+- Maintenance burden: as verifications of an op are contained in two different methods.
+
+The solution is as follows:
+
+1. **For most ops w/o regions** (like PadOp):
+Put all the verification code into the shape functions, and discard verifiers totally.
+
+2. **For ops w/ regions** (like `ReduceOp/IfOp`, a full list is [here](https://github.com/openxla/stablehlo/pull/401)): the `builder()` of the op can be created before all the components are in place, for [example](https://github.com/tensorflow/mlir-hlo/blob/master/mhlo/transforms/mhlo_canonicalize_reduction/mhlo_canonicalize_reduction.cc#L221), the `ReduceOp` is created without regions and soon the shape functions are used.
+    1. If the regions are not needed for type inference (like `ReduceOp`), put the region related verification logic in verifiers. Duplicate some code if it is inevitable.
+    2. If the regions are needed for type inference (`IfOp/CaseOp/MapOp`), the shape function must verify the regions are not empty explicitly, even though the ODS may already guarantee its existence in the Op definition.
```

Comment:
"If the regions are needed for type inference (`IfOp/CaseOp/MapOp`), then additionally the shape function...".